cpp/include/rapidsmpf/streaming/core/spillable_messages.hpp

```diff
@@ -0,0 +1,132 @@
+ * spilled or extracted independently. Each message is assigned a unique
+ * `MessageId` upon insertion, which can later be used to extract or spill that
+ * message.
+ *
+ * The container is thread-safe for concurrent insertions, extractions, and
+ * spills.
+ *
+ * ### Example
+ * @code{.cpp}
+ * SpillableMessages messages;
+ * auto id = messages.insert(to_message(TableChunk{...}));
+ * messages.spill(id, ...);
+ * auto msg = messages.extract(id);
+ * @endcode
+ */
+class SpillableMessages {
+  public:
+    /// @brief Unique identifier assigned to each message.
+    using MessageId = std::uint64_t;
+
+    SpillableMessages() = default;
+    SpillableMessages(SpillableMessages const&) = delete;
+    SpillableMessages& operator=(SpillableMessages const&) = delete;
+    SpillableMessages(SpillableMessages&&) noexcept = delete;
+    SpillableMessages& operator=(SpillableMessages&&) noexcept = delete;
+
+    /**
+     * @brief Insert a new message and return its assigned ID.
+     *
+     * @param message Message to insert.
+     * @return Assigned `MessageId` of the inserted message.
+     */
+    [[nodiscard]] MessageId insert(Message&& message);
+
+    /**
+     * @brief Extract and remove a message by ID.
+     *
+     * If the message is currently being spilled, this method blocks until spilling
+     * completes.
+     *
+     * @param mid Message identifier.
+     * @return Extracted `Message` instance.
+     *
+     * @throws std::out_of_range If the message ID does not exist.
+     */
+    [[nodiscard]] Message extract(MessageId mid);
+
+    /**
+     * @brief Spill a message's device memory to host memory.
+     *
+     * Performs an in-place deep copy of the message's payload from device to
+     * host memory using the specified buffer resource.
+     *
+     * If the message is currently being accessed by another thread, is already
+     * spilled, not spillable, or does not exist, the operation returns immediately
+     * without spilling.
+     *
+     * @param mid Message identifier. If the message does not exist, zero is returned.
+     * @param br Buffer resource used for allocations during the spill operation.
+     * @return Number of bytes released from device memory (0 if nothing was spilled).
+     */
+    [[nodiscard]] std::size_t spill(MessageId mid, BufferResource* br) const;
+
+    /**
+     * @brief Get a snapshot of current messages' content descriptions.
+     *
+     * The returned map may become outdated immediately if other threads
+     * modify the container after this call.
+     *
+     * Use this snapshot to decide which messages to spill, but keep in mind
+     * that the information may no longer be accurate when the actual spill
+     * occurs. When calling `spill()`, the returned size reflects what was
+     * actually spilled.
+     *
+     * @return Copy of a map from `MessageId` to `ContentDescription`.
+     */
+    std::map<MessageId, ContentDescription> get_content_descriptions() const;
```

Comment:
Not sure if that would be useful, but would a method that does `spill` also immediately be helpful? For example, pass a function as argument that executes looking for the "perfect match" (e.g., largest message) and immediately does the spilling without risking invalidation of the returned value of `get_content_descriptions()`?